engineapi/engineapi/middleware.py

```diff
@@ -26,13 +33,170 @@
+    """
+
+
+class BugoutUnverifiedAuth(Exception):
+    """
+    Raised when attempted access by unverified Brood account.
+    """
+
+
+class BugoutAuthWrongApp(Exception):
+    """
+    Raised when user does not belong to this application.
+    """
+
+
+def parse_auth_header(auth_header: str) -> Tuple[str, str]:
+    """
+    Returns: auth_format and user_token passed in authorization header.
+    """
+    auth_list = auth_header.split()
+    if len(auth_list) != 2:
+        raise InvalidAuthHeaderFormat("Wrong authorization header")
+
+    return auth_list[0], auth_list[1]
+
+
+def bugout_auth(token: str) -> BugoutUser:
+    """
+    Extended bugout.get_user with additional checks.
+    """
+    user: BugoutUser = bc.get_user(token)
+    if not user.verified:
+        raise BugoutUnverifiedAuth("Only verified accounts can have access")
+    if str(user.application_id) != str(MOONSTREAM_APPLICATION_ID):
+        raise BugoutAuthWrongApp("User does not belong to this application")
+
+    return user
+
+
+def brood_auth(token: UUID) -> BugoutUser:
+    try:
+        user: BugoutUser = bugout_auth(token=token)
+    except BugoutUnverifiedAuth:
+        logger.info(f"Attempted access by unverified Brood account: {user.id}")
+        raise EngineHTTPException(
+            status_code=403,
+            detail="Only verified accounts can have access",
+        )
+    except BugoutAuthWrongApp:
+        raise EngineHTTPException(
+            status_code=403,
+            detail="User does not belong to this application",
+        )
+    except BugoutResponseException as e:
+        raise EngineHTTPException(
+            status_code=e.status_code,
+            detail=e.detail,
+        )
+    except Exception as e:
+        logger.error(f"Error processing Brood response: {str(e)}")
+        raise EngineHTTPException(
+            status_code=500,
+            detail="Internal server error",
+        )
+
+    return user
+
+
+async def request_user_auth(
+    token: UUID = Depends(oauth2_scheme),
+) -> BugoutUser:
+    user = brood_auth(token=token)
+
+    return user
+
+
+async def request_none_or_user_auth(
+    authorization: str = Header(None),
+) -> Optional[BugoutUser]:
+    """
+    Fetch Bugout user if authorization token provided.
+    """
+    user: Optional[BugoutUser] = None
+    if authorization is not None:
+        token: str = ""
+        try:
+            _, token = parse_auth_header(auth_header=authorization)
+        except InvalidAuthHeaderFormat:
+            raise EngineHTTPException(
+                status_code=403, detail="Wrong authorization header"
+            )
+        except Exception as e:
+            logger.error(f"Error parsing auth header: {str(e)}")
+            raise EngineHTTPException(status_code=500, detail="Internal server error")
+
+        if token != "":
+            user = brood_auth(token=token)
+
+    return user
+
+
+async def metatx_sign_header(
```

Comment:
`metatx_sign_header` is a confusing name - it seems like it is *signing* something. But actually, it is *verifying* a signature.

Maybe `metatx_verify_header`?